appinventor/appengine/src/com/google/appinventor/server/GallerySearchIndex.java

```diff
@@ -113,32 +113,43 @@ public GalleryAppListResult find (String searchWords, int start, int count) {
     final List<GalleryApp> apps = new ArrayList<GalleryApp>();
     final Result<Integer> size = new Result<Integer>();
 
-    try {
-      //New search
-      if (start == 0 || searchCursor == null){
-        searchCursor = Cursor.newBuilder().build();
+    //New search
+    if (start == 0 || searchCursor == null){
+      searchCursor = Cursor.newBuilder().build();
+    }
+    Query query = Query.newBuilder()
+            .setOptions(QueryOptions.newBuilder()
+                    .setCursor(searchCursor)
+                    .setLimit(count)
+                    .setNumberFoundAccuracy(appResultMaxAccuracy)
+                    // for deployed apps, uncomment the line below to demo snippeting.
+                    // This will not work on the dev_appserver.
+                    // setFieldsToSnippet("content").
+                    .build())
+            .build(searchWords);
+
+    Results<ScoredDocument> results = null;
+    int attempts = 0;
+    while (attempts++ < 3){
+      try {
+        LOG.info("Sending query " + query);
+        results = getIndex().search(query);
+        break;
+      } catch (SearchException e) {
+        if (StatusCode.TRANSIENT_ERROR.equals(e.getOperationResult().getCode())) {
+          // retry
+          LOG.info("Query failed on attempt:"+ attempts + ". retrying...");
```

Comment:
@toropeza This doesn't seem correct. If the error is NOT a TRANSIENT_ERROR, then you will still retry, but just be silent. Errors that are not TRANSIENT should cause a failure without retrying...